File: langs/zig/primes/Atkin.zig
const std = @import("std");

const UPPER_BOUND = 5000000;
const PREFIX = 32338;

const Node = struct {
    children: std.AutoHashMap(u8, *Node),
    terminal: bool,
    allocator: std.mem.Allocator,

    fn init(allocator: std.mem.Allocator) !*Node {
        const node = try allocator.create(Node);
        node.* = Node{
            .children = std.AutoHashMap(u8, *Node).init(allocator),
            .terminal = false,
            .allocator = allocator,
        };
        return node;
    }

    fn deinit(self: *Node) void {
        var it = self.children.iterator();
        while (it.next()) |entry| {
            entry.value_ptr.*.deinit();
        }
        self.children.deinit();
        self.allocator.destroy(self);
    }
};

const Sieve = struct {
    limit: i32,
    prime: []bool,
    allocator: std.mem.Allocator,

    fn init(allocator: std.mem.Allocator, limit: i32) !Sieve {
        const prime = try allocator.alloc(bool, @intCast(limit + 1));
        @memset(prime, false);
        return Sieve{
            .limit = limit,
            .prime = prime,
            .allocator = allocator,
        };
    }

    fn deinit(self: *Sieve) void {
        self.allocator.free(self.prime);
    }

    fn toList(self: *const Sieve, allocator: std.mem.Allocator) !std.ArrayList(i32) {
        var result = std.ArrayList(i32).init(allocator);
        try result.append(2);
        try result.append(3);

        var p: i32 = 5;
        while (p <= self.limit) : (p += 1) {
            if (self.prime[@intCast(p)]) {
                try result.append(p);
            }
        }

        return result;
    }

    fn omitSquares(self: *Sieve) *Sieve {
        var r: i32 = 5;
        while (r * r < self.limit) : (r += 1) {
            if (self.prime[@intCast(r)]) {
                var i: i32 = r * r;
                while (i < self.limit) : (i += r * r) {
                    self.prime[@intCast(i)] = false;
                }
            }
        }
        return self;
    }

    fn step1(self: *Sieve, x: i32, y: i32) void {
        const n = (4 * x * x) + (y * y);
        if (n <= self.limit and (n % 12 == 1 or n % 12 == 5)) {
            self.prime[@intCast(n)] = !self.prime[@intCast(n)];
        }
    }

    fn step2(self: *Sieve, x: i32, y: i32) void {
        const n = (3 * x * x) + (y * y);
        if (n <= self.limit and n % 12 == 7) {
            self.prime[@intCast(n)] = !self.prime[@intCast(n)];
        }
    }

    fn step3(self: *Sieve, x: i32, y: i32) void {
        const n = (3 * x * x) - (y * y);
        if (x > y and n <= self.limit and n % 12 == 11) {
            self.prime[@intCast(n)] = !self.prime[@intCast(n)];
        }
    }

    fn loopY(self: *Sieve, x: i32) void {
        var y: i32 = 1;
        while (y * y < self.limit) : (y += 1) {
            self.step1(x, y);
            self.step2(x, y);
            self.step3(x, y);
        }
    }

    fn loopX(self: *Sieve) void {
        var x: i32 = 1;
        while (x * x < self.limit) : (x += 1) {
            self.loopY(x);
        }
    }

    fn calc(self: *Sieve) *Sieve {
        self.loopX();
        return self.omitSquares();
    }
};

fn generateTree(allocator: std.mem.Allocator, l: std.ArrayList(i32)) !*Node {
    const root = try Node.init(allocator);

    for (l.items) |el| {
        var head = root;
        
        var buf: [20]u8 = undefined;
        const str = try std.fmt.bufPrint(&buf, "{d}", .{el});
        
        for (str) |ch| {
            if (!head.children.contains(ch)) {
                const new_node = try Node.init(allocator);
                try head.children.put(ch, new_node);
            }
            head = head.children.get(ch).?;
        }
        
        head.terminal = true;
    }

    return root;
}

const QueueItem = struct {
    node: *Node,
    prefix: []const u8,
};

fn find(allocator: std.mem.Allocator, upper_bound: i32, prefix: i32) !std.ArrayList(i32) {
    var sieve = try Sieve.init(allocator, upper_bound);
    defer sieve.deinit();
    
    _ = sieve.calc();
    const primes_list = try sieve.toList(allocator);
    defer primes_list.deinit();

    var buf: [20]u8 = undefined;
    const str_prefix = try std.fmt.bufPrint(&buf, "{d}", .{prefix});
    
    const root = try generateTree(allocator, primes_list);
    defer root.deinit();

    var head = root;
    for (str_prefix) |ch| {
        if (head.children.get(ch)) |child| {
            head = child;
        } else {
            return std.ArrayList(i32).init(allocator);
        }
    }

    var queue = std.ArrayList(QueueItem).init(allocator);
    defer {
        for (queue.items) |item| {
            allocator.free(item.prefix);
        }
        queue.deinit();
    }

    const initial_prefix = try allocator.dupe(u8, str_prefix);
    try queue.append(QueueItem{ .node = head, .prefix = initial_prefix });

    var result = std.ArrayList(i32).init(allocator);

    while (queue.items.len > 0) {
        const item = queue.pop();
        const top = item.node;
        const prefix_str = item.prefix;
        defer allocator.free(prefix_str);

        if (top.terminal) {
            const num = try std.fmt.parseInt(i32, prefix_str, 10);
            try result.append(num);
        }

        var it = top.children.iterator();
        while (it.next()) |entry| {
            const ch = entry.key_ptr.*;
            const v = entry.value_ptr.*;
            
            const new_prefix = try std.fmt.allocPrint(allocator, "{s}{c}", .{ prefix_str, ch });
            try queue.insert(0, QueueItem{ .node = v, .prefix = new_prefix });
        }
    }

    return result;
}

pub fn main() anyerror!void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const stdout = std.io.getStdOut().writer();

    const start_time = std.time.milliTimestamp();

    const result = try find(allocator, UPPER_BOUND, PREFIX);
    defer result.deinit();

    try stdout.print("[", .{});
    for (result.items, 0..) |item, i| {
        if (i > 0) {
            try stdout.print(", ", .{});
        }
        try stdout.print("{d}", .{item});
    }
    try stdout.print("]\n", .{});

    const end_time = std.time.milliTimestamp();
    const duration = end_time - start_time;

    try stdout.print("Execution time: {d}ms\n", .{duration});
}

File: langs/c/primes/Atkin.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <time.h>

#define UPPER_BOUND 5000000
#define PREFIX 32338

// -----------------------------
// Struct definitions
// -----------------------------

typedef struct Node {
    struct ChildMap *children;
    int num_children;
    bool terminal;
} Node;

typedef struct ChildMap {
    char key;
    struct Node *value;
} ChildMap;

typedef struct {
    int limit;
    bool *prime;
} Sieve;

// -----------------------------
// Function declarations
// -----------------------------
Sieve* Sieve_new(int limit);
void Sieve_free(Sieve *self);
Sieve* Sieve_calc(Sieve *self);
Sieve* Sieve_omit_squares(Sieve *self);
void step1(Sieve *self, int x, int y);
void step2(Sieve *self, int x, int y);
void step3(Sieve *self, int x, int y);
void loop_y(Sieve *self, int x);
void loop_x(Sieve *self);
int* to_list(Sieve *self, int *out_size);

Node* Node_new();
void Node_free(Node *node);
Node* generate_tree(int *list, int size);
Node* Node_get_child(Node *node, char ch);
Node* Node_add_child(Node *node, char ch);

int* find(int upper_bound, int prefix, int *out_size);
char* to_string(int *arr, int size);

// -----------------------------
// Sieve Implementation
// -----------------------------

Sieve* Sieve_new(int limit) {
    Sieve *self = malloc(sizeof(Sieve));
    self->limit = limit;
    self->prime = calloc(limit + 1, sizeof(bool));
    return self;
}

void Sieve_free(Sieve *self) {
    if (!self) return;
    free(self->prime);
    free(self);
}

Sieve* Sieve_omit_squares(Sieve *self) {
    for (int r = 5; r * r < self->limit; ++r) {
        if (self->prime[r]) {
            for (int i = r * r; i < self->limit; i += r * r) {
                self->prime[i] = false;
            }
        }
    }
    return self;
}

void step1(Sieve *self, int x, int y) {
    int n = (4 * x * x) + (y * y);
    if (n <= self->limit && (n % 12 == 1 || n % 12 == 5)) {
        self->prime[n] = !self->prime[n];
    }
}

void step2(Sieve *self, int x, int y) {
    int n = (3 * x * x) + (y * y);
    if (n <= self->limit && n % 12 == 7) {
        self->prime[n] = !self->prime[n];
    }
}

void step3(Sieve *self, int x, int y) {
    int n = (3 * x * x) - (y * y);
    if (x > y && n <= self->limit && n % 12 == 11) {
        self->prime[n] = !self->prime[n];
    }
}

void loop_y(Sieve *self, int x) {
    for (int y = 1; y * y < self->limit; ++y) {
        step1(self, x, y);
        step2(self, x, y);
        step3(self, x, y);
    }
}

void loop_x(Sieve *self) {
    for (int x = 1; x * x < self->limit; ++x) {
        loop_y(self, x);
    }
}

Sieve* Sieve_calc(Sieve *self) {
    loop_x(self);
    return Sieve_omit_squares(self);
}

int* to_list(Sieve *self, int *out_size) {
    int *result = malloc((self->limit + 1) * sizeof(int));
    int idx = 0;
    result[idx++] = 2;
    result[idx++] = 3;
    for (int p = 5; p <= self->limit; ++p) {
        if (self->prime[p]) {
            result[idx++] = p;
        }
    }
    *out_size = idx;
    return result;
}

// -----------------------------
// Trie Tree (Node) Implementation
// -----------------------------

Node* Node_new() {
    Node *node = malloc(sizeof(Node));
    node->children = NULL;
    node->num_children = 0;
    node->terminal = false;
    return node;
}

Node* Node_get_child(Node *node, char ch) {
    for (int i = 0; i < node->num_children; ++i) {
        if (node->children[i].key == ch) {
            return node->children[i].value;
        }
    }
    return NULL;
}

Node* Node_add_child(Node *node, char ch) {
    node->children = realloc(node->children, (node->num_children + 1) * sizeof(ChildMap));
    node->children[node->num_children].key = ch;
    node->children[node->num_children].value = Node_new();
    node->num_children++;
    return node->children[node->num_children - 1].value;
}

void Node_free(Node *node) {
    if (!node) return;
    for (int i = 0; i < node->num_children; ++i) {
        Node_free(node->children[i].value);
    }
    free(node->children);
    free(node);
}

Node* generate_tree(int *list, int size) {
    Node *root = Node_new();
    char buffer[32];
    for (int i = 0; i < size; ++i) {
        snprintf(buffer, sizeof(buffer), "%d", list[i]);
        Node *head = root;
        for (int j = 0; j < strlen(buffer); ++j) {
            Node *child = Node_get_child(head, buffer[j]);
            if (!child) {
                child = Node_add_child(head, buffer[j]);
            }
            head = child;
        }
        head->terminal = true;
    }
    return root;
}

// -----------------------------
// BFS Find Function
// -----------------------------

typedef struct QueueNode {
    Node *node;
    char *prefix;
    struct QueueNode *next;
} QueueNode;

void enqueue(QueueNode **front, QueueNode **rear, Node *node, const char *prefix) {
    QueueNode *new_node = malloc(sizeof(QueueNode));
    new_node->node = node;
    new_node->prefix = strdup(prefix);
    new_node->next = NULL;
    if (*rear) (*rear)->next = new_node;
    else *front = new_node;
    *rear = new_node;
}

QueueNode* dequeue(QueueNode **front, QueueNode **rear) {
    if (!*front) return NULL;
    QueueNode *temp = *front;
    *front = (*front)->next;
    if (!*front) *rear = NULL;
    return temp;
}

int* find(int upper_bound, int prefix, int *out_size) {
    Sieve *s = Sieve_new(upper_bound);
    Sieve_calc(s);
    int prime_count;
    int *primes = to_list(s, &prime_count);

    char str_prefix[32];
    snprintf(str_prefix, sizeof(str_prefix), "%d", prefix);

    Node *head = generate_tree(primes, prime_count);
    for (int i = 0; i < strlen(str_prefix); ++i) {
        head = Node_get_child(head, str_prefix[i]);
        if (!head) {
            *out_size = 0;
            free(primes);
            Sieve_free(s);
            return NULL;
        }
    }

    QueueNode *front = NULL, *rear = NULL;
    enqueue(&front, &rear, head, str_prefix);

    int *result = malloc(1024 * sizeof(int));
    int count = 0, capacity = 1024;

    while (front) {
        QueueNode *q = dequeue(&front, &rear);
        Node *node = q->node;

        if (node->terminal) {
            if (count >= capacity) {
                capacity *= 2;
                result = realloc(result, capacity * sizeof(int));
            }
            result[count++] = atoi(q->prefix);
        }

        for (int i = 0; i < node->num_children; ++i) {
            char new_prefix[64];
            snprintf(new_prefix, sizeof(new_prefix), "%s%c", q->prefix, node->children[i].key);
            enqueue(&front, &rear, node->children[i].value, new_prefix);
        }

        free(q->prefix);
        free(q);
    }

    *out_size = count;

    Node_free(head);
    free(primes);
    Sieve_free(s);

    return result;
}

// -----------------------------
// Utility Functions
// -----------------------------

char* to_string(int *arr, int size) {
    char *buf = malloc(size * 12);
    buf[0] = '\0';
    strcat(buf, "[");
    for (int i = 0; i < size; ++i) {
        char num[16];
        sprintf(num, "%d", arr[i]);
        strcat(buf, num);
        if (i < size - 1) strcat(buf, ", ");
    }
    strcat(buf, "]");
    return buf;
}

// -----------------------------
// Main Function
// -----------------------------

int main() {
    clock_t start_time = clock();

    int result_size;
    int *result = find(UPPER_BOUND, PREFIX, &result_size);

    char *result_str = to_string(result, result_size);
    printf("%s\n", result_str);

    clock_t end_time = clock();
    double duration = ((double)(end_time - start_time) / CLOCKS_PER_SEC) * 1000.0;
    printf("Execution time: %.2fms\n", duration);

    free(result);
    free(result_str);
    return 0;
}
